### PR TITLE
Fix 9332

### LINKED ISSUE #9332: Global values defined in a subchart can affect global values in parent chart and sibling charts
The issue can be reproduced with a chart that contains 2 subcharts and the following values:

1. toplevel values:
    ```
    global:
      nested:
        v1: chart
    ```

2. values in subchart alpha:
    ```
    global:
      nested:
        v2: alpha
        v3: alpha
    ```
3. values in subchart beta:
    ```
    global:
      nested:
        v2: beta
        v4: beta
    ```

After installing such a chart, `helm get values -a` produced the following output:

```
COMPUTED VALUES:
alpha:
  global:
    nested:
      v1: chart
      v2: alpha
      v3: alpha
      v4: beta
beta:
  global:
    nested:
      v1: chart
      v2: alpha
      v3: alpha
      v4: beta
global:
  nested:
    v1: chart
    v2: alpha
    v3: alpha
    v4: beta

```

This shows that values from the subcharts become visible in the toplevel chart and in other subcharts,

In my opinion, this behavior contradicts the following statement, which can be found in https://helm.sh/docs/topics/charts/:
> "There is no way for a subchart to influence the values of the parent chart."

Here is also a zip file that contains my example chart to reproduce the issue: [chart.zip](https://github.com/helm/helm/files/5933020/chart.zip)

Output of `helm version`:
```
version.BuildInfo{Version:"v3.3.0", GitCommit:"8a4aeec08d67a7b84472007529e8097ec3742105", GitTreeState:"dirty", GoVersion:"go1.14.7"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:23:04Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): N/A

### PR DESCRIPTION
**Pull request by bgutjahr**
**What this PR does / why we need it**:

This PR fixes issue 9332, which is about global values defined in subchart can affect globals in parent chart or sibling charts.

**Special notes for your reviewer**:

The fix consists of 3 commits.

- a unit test enhancement, which failed before the fix was implemented.
- the actual fix, which was just a missing 'continue' statement, such that the intended code was later overrule
- an additional code improvement of `coalesce.go to` get rid or continue statement and using appropriate else
  blocks instead. This is not actually necessary for the fix, but should improve the code to make a similar issue
  less likely.

**If applicable**:
- [X] this PR contains unit tests

closes #9332

### CODE REVIEW

**Review by adamreese — APPROVED** [review]
I approve this mostly based on the unit tests.  We should get a second review because this section of code produces a lot of bugs.

**Comment by bacongobbler** [comment]
Scheduling for 3.7.0 so we don’t miss this during review.

Can you please rebase this?

**Comment by bgutjahr** [comment]
After rebasing, I found that the defect has already been fixed end of May 2021 with commit 3b68afc1c820d456e8281a6e6911af59c4b661c8 on the main branch. So I assume that fix will already be part of 3.7.0.

What remains now are two commits in the PR. One with additional coalesce test cases, and the other with a code approval in coalesce.go by getting rid of continue statements.

Feel free to merge these improvements, I you find them valuable. Since the actual defect is fixed, I won't prolong the PR again after the next 90 days expiration.

**Comment by mattfarina** [comment]
@bgutjahr can you please fix the failing test.

**Review by hickeyma — APPROVED** [review]
LGTM, even though this was fixed in #9756, it is worth adding the unit tests and the cleanup of the deprecated continue statements. Thank for working on this @bgutjahr

**Pull request merged**
